### PR TITLE
Support Unix domain socket (UDS) server

### LINKED ISSUE #25: [Server] Add support for Unix domain sockets (UDS)

### PR DESCRIPTION
**Pull request by andig**
~~~Draft implementation for https://github.com/reactphp/socket/issues/25, tests still missing~~~

Resolves / closes #25

### CODE REVIEW

**Comment by andig** [comment]
> Can you  add the missing tests and documentation in the README file? Also, this should probably be integrated into the main Server class.

Done :)

File: src/UnixServer.php [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+     * @throws RuntimeException if listening on this address fails (already in use etc.)
+     */
+    public function __construct($path, LoopInterface $loop, array $context = array())
+    {
+        $this->loop = $loop;
+
+        if (strpos($path, '://') === false) {
+            $path = 'unix://' . $path;
+        } elseif (substr($path, 0, 7) !== 'unix://') {
+            throw new \InvalidArgumentException('Given URI "' . $path . '" is InvalidArgumentException');
```
Comment (jsor) [inline src/UnixServer.php]: `...is InvalidArgumentException` -> `...is invalid`

**Comment by clue** [comment]
Thank you for the quick update!

I don't see any major issues after eyeballing this and would love to get this in :shipit: I'll try to get back to this tomorrow and address some minor outstanding formatting issues myself :+1:

File: README.md [hunk 2]
```diff
@@ -373,6 +374,12 @@ brackets:
 $server = new Server('[::1]:8080', $loop);
 ```
 
+To listen on a Unix domain socket, the uri MUST be prefixed with the `unix://` scheme:
+
+```php
+$server = new Server('unix:///tmp/.sock', $loop);
```
Comment (WyriHaximus) [inline README.md]: :heart:

File: tests/UnixServerTest.php [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+            $server->close();
+        });
+
+        $this->loop->run();
+
+        $this->assertEquals($bytes, $received);
+    }
+
+    /**
+     * @covers React\EventLoop\StreamSelectLoop::tick
```
Comment (WyriHaximus) [inline tests/UnixServerTest.php]: Why does this test covers `React\EventLoop\StreamSelectLoop::tick`?

File: tests/UnixServerTest.php [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+
+        $this->server->on('connection', function ($conn) use ($mock) {
+            $conn->on('end', $mock);
+        });
+        $this->loop->tick();
+        $this->loop->tick();
+    }
+
+    /**
+     * @covers React\EventLoop\StreamSelectLoop::tick
```
Comment (WyriHaximus) [inline tests/UnixServerTest.php]: Why does this test covers `React\EventLoop\StreamSelectLoop::tick`?

File: tests/UnixServerTest.php [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+    public function tearDown()
+    {
+        if ($this->server) {
+            $this->server->close();
+        }
+    }
+
+    private function getRandomSocketUri()
+    {
+        return "unix://" . sys_get_temp_dir() . DIRECTORY_SEPARATOR . uniqid(rand(), true) . '.sock';
```
Comment (WyriHaximus) [inline tests/UnixServerTest.php]: Just nitpicking here but why not go with `uniqid('reactphp-socket-tests-unix-server-', true)` to make it explicit when debugging that it's a socket for the `UnixServer` test?

File: tests/UnixServerTest.php (hunk 4 above)
Comment (andig) [reply tests/UnixServerTest.php]: I must admit that the tests are not original art but all copied from `TcpServer`- so that's a copy/paste artefact. Remove the covers hint here?

Comment (WyriHaximus) [reply tests/UnixServerTest.php]: Yeah noticed that, but to be honest that is alright. There are few distinct differences that I'm not bothered by it 😄 . The hints are fine as long as they make sense, covering a loop tick in a server test doesn't =D

File: tests/UnixServerTest.php (hunk 5 above)
Comment (andig) [reply tests/UnixServerTest.php]: Turned out to not be a good idea, see https://github.com/reactphp/socket/pull/120/commits/999376385b95f3fd70bdc757cea006821e2dd020. I've added a check for the client as well.

File: tests/UnixServerTest.php (hunk 4 above)
Comment (andig) [reply tests/UnixServerTest.php]: Should I change the TcpServerTest, too?

Comment (WyriHaximus) [reply tests/UnixServerTest.php]: Nah out of scope for this PR

File: tests/UnixServerTest.php (hunk 5 above)
Comment (WyriHaximus) [reply tests/UnixServerTest.php]: Interesting, what happened?

Comment (andig) [reply tests/UnixServerTest.php]: Warning is shown (in the tests without @) that the path is truncated. My thinking was that we should rather be explicit with failure instead of the user potentially reusing an existing socket due to same name.

File: src/UnixConnector.php [hunk 6]
```diff
@@ -31,8 +31,9 @@ public function connect($path)
             return Promise\reject(new InvalidArgumentException('Given URI "' . $path . '" is invalid'));
         }
 
-        if (strlen($path) > 104) {
-            return Promise\reject(new InvalidArgumentException('Given URI "' . $path . '" exceeds maximum allowed length of 104 bytes'));
+        $maxLen = PHP_OS === 'Darwin' ? 104 : 108;
```
Comment (kelunik) [inline src/UnixConnector.php]: Is the max-length including or excluding the trailing NUL byte? A link would probably be good for future reference.

File: tests/UnixServerTest.php [hunk 7]
```diff
@@ -268,8 +268,8 @@ public function testListenOnBusyPortThrows()
      */
     public function testConnectWithOverlyLongAddress()
     {
-        // string > 104 characters
-        $path = "unix://" . sys_get_temp_dir() . DIRECTORY_SEPARATOR . str_repeat('_', 100) . '.sock';
+        // string > 104/108 characters
+        $path = "unix://" . sys_get_temp_dir() . DIRECTORY_SEPARATOR . str_repeat('_', 104) . '.sock';
```
Comment (kelunik) [inline tests/UnixServerTest.php]: It should probably subtract the length of the `sys_get_temp_dir()` for the `str_repeat`.

Comment (andig) [reply tests/UnixServerTest.php]: It just needs to be long enough to trigger the exception? I did not want to go into complex scenarios here.

Comment (kelunik) [reply tests/UnixServerTest.php]: Yes, but it should trigger on exactly X byte and succeed otherwise, no? Probably should have a tests that it works with one less character.

Comment (clue) [reply tests/UnixServerTest.php]: Thanks for the discussion so far! I wonder if we really need this check in the first place? It looks like `stream_socket_server()` rejects an invalid path anyway, so why not rely on this?

https://3v4l.org/Gff8D

This would allow us to remove any platform specific constants / assumptions. What do you think about this?

Comment (clue) [reply tests/UnixServerTest.php]: Never mind, it looks like it actually still succeeds despite this warning and instead creates a truncated socket path. Sorry for the confusion, go ahead :-)

File: tests/UnixConnectorTest.php [hunk 8]
```diff
@@ -65,12 +65,28 @@ public function testValid()
 
     public function testConnectWithOverlyLongAddress()
     {
+        https://serverfault.com/questions/641347/check-if-a-path-exceeds-maximum-for-unix-domain-socket
```
Comment (kelunik) [inline tests/UnixConnectorTest.php]: Missing `//`.

File: tests/UnixServerTest.php (hunk 7 above)
Comment (andig) [reply tests/UnixServerTest.php]: right, your test is invalid, see https://3v4l.org/4OPt0

**Comment by the-eater** [comment]
What are currently the blocking issues with this PR? need it in my project and would like to get rid of the `dev-unix as v0.8.4`, also wouldn't it be sane to stop supporting PHP5.3-5.5 as they are all EOL?

**Comment by andig** [comment]
I will remove the loop->tick from the tests, then it should imho be ready for final review

File: src/Server.php [hunk 9]
```diff
@@ -28,10 +29,15 @@ public function __construct($uri, LoopInterface $loop, array $context = array())
...
+            $server = new UnixServer($uri, $loop, $context['unix']);
+        }
+        else {
+            $server = new TcpServer(str_replace('tls://', '', $uri), $loop, $context['tcp']);
 
-        if ($scheme === 'tls') {
-            $server = new SecureServer($server, $loop, $context['tls']);
+            if ($scheme === 'tls') {
+                $server = new SecureServer($server, $loop, $context['tls']);
+            }
```
Comment (the-eater) [inline src/Server.php]: Maybe export the Server constructing logic to a separate method like this: https://github.com/the-eater/socket/commit/2defb1b146f85dcd28510d725ce7b7aece7c644d?diff=split

**Comment by andig** [comment]
Ping @WyriHaximus what is missing to get this aging pr merged? Final comment above?

**Comment by WyriHaximus** [comment]
@andig I will have a look this weekend but last time I ran into some issues. Will ping you with some questions then 👍

**Comment by clue** [comment]
Thank you @andig! We're currently trying to focus our efforts on stabilizing things so we can get a stable v1.0.0 release out as soon as possible. This features was originally planned for v1.1.0 via #25, but given the high quality of this PR I'm open to getting this in before the v1.0.0 release. :shipit: 

I would like to get the outstanding v0.8.5 bug fix release out in the next days first before introducing this new feature. Does it make sense for you to postpone this to the next v0.8.6 release so we can look into this immediately after the bug fix release is out?

**Comment by andig** [comment]
@clue I was hoping to get this into 0.8.5. It is a new feature but does not break compatibility. Only exception is the handling of maximum UDS path lengths that would otherwise potentially have caused unexpected behaviour. So- the sooner the better as ppm depends on it.

File: src/UnixConnector.php [hunk 10]
```diff
@@ -27,7 +28,12 @@ public function connect($path)
         if (strpos($path, '://') === false) {
             $path = 'unix://' . $path;
         } elseif (substr($path, 0, 7) !== 'unix://') {
-            return Promise\reject(new \InvalidArgumentException('Given URI "' . $path . '" is invalid'));
+            return Promise\reject(new InvalidArgumentException('Given URI "' . $path . '" is invalid'));
+        }
+
+        $maxLen = PHP_OS === 'Darwin' ? 104 : 108;
+        if (strlen($path) > $maxLen) {
+            return Promise\reject(new InvalidArgumentException('Given URI "' . $path . '" exceeds maximum allowed length'));
```
Comment (clue) [inline src/UnixConnector.php]: I think you're addressing a relevant issue here, thank you! However, I would both get a better understanding of this problem before getting this in and also believe that this is unrelated to this new feature. Does it make sense to you to remove this for now and file this a separate PR? :+1: 

Some thoughts (short brain dump only):
* The path limit seems to apply to the path as given, a prior `chdir()` may be used to work around this for use use cases
* The `unix://` scheme is not included in this limit
* The limit is system dependent and 108 for Linux and 104 for BSD (Mac) and lower for other systems
* Linux man pages suggest `108` is a valid length, but PHP already raises a notice here
* The E_NOTICE was apparently introduced via https://bugs.php.net/bug.php?id=60106 in PHP 5.4.1+
* All relevant versions (including < PHP 5.4.1) already (silently) truncated the path
* My experiments: https://3v4l.org/hSRCU

**Comment by WyriHaximus** [comment]
@andig could you share your nginx config? And how do you deal with existing `.sock` files when you haven't started the process yet?

**Comment by andig** [comment]
@clue thank you for the review. I have removed all path length stuff and will raise WIP PR once this is in and I can rebase on top of it.

> could you share your nginx config? 

@WyriHaximus I can't- no nginx in use, sorry.

> And how do you deal with existing .sock files when you haven't started the process yet?

Not sure. If the socket is not in use the server will allow this. What is the expected behaviour?

File: src/UnixConnector.php (hunk 10 above)
Comment (andig) [reply src/UnixConnector.php]: Thanks for the experiments :)

> The path limit seems to apply to the path as given, a prior chdir() may be used to work around this for use use cases

Would need to double-check.

> The unix:// scheme is not included in this limit

Why do you think so? The previous unit tests seemed to indicate that it does?

> The limit is system dependent and 108 for Linux and 104 for BSD (Mac) and lower for other systems

Unless we could identify from some setting which those are we could only live with assumptions. Custom kernels might have yet other limits?

> Linux man pages suggest 108 is a valid length, but PHP already raises a notice here

I can only imagine PHP takes the trailing 0 into account for that check. Might be a bug?

> All relevant versions (including < PHP 5.4.1) already (silently) truncated the path

Meaning your suggestion would be to accept this situation? 

I'd be happy to not re-raise this PR but thought it would be good as a precaution as I've only accidentally hit this limit and it might have gone unnoticed.

Long story short: do you think it is worth pursuing this or rather entirely skip it?

File: src/Server.php (hunk 9 above)
Comment (clue) [reply src/Server.php]: Functionally LGTM, however I don't currently see this covered in the test suite (plus minor CS issue above). Can you update this? :shipit: :+1:

Comment (andig) [reply src/Server.php]: @clue what's your expectation here? Testing the type of the create private server member? Would need a hint how to do this due to the private nature.

Comment (clue) [reply src/Server.php]: This class is a facade that mostly contains some glue code to make the underlying APIs more easily accessible. Just check the other high-level functional tests and the code coverage. IMHO it's sufficient to create an Unix domain socket server with this facade and check a client connection can be established.

Comment (andig) [reply src/Server.php]: Gotcha. Tests added.

**Comment by andig** [comment]
Will check the test failures, but those are unrelated.

**Comment by andig** [comment]
Should the OSX test failures re-appear please check https://github.com/travis-ci/travis-ci/issues/8552#issuecomment-335430374

**Comment by clue** [comment]
Unrelated Mac OS X test failure has been resolved by restarting build, see also #134 for follow-up PR :shipit:

**Comment by andig** [comment]
@clue is this something that could get packaged into `0.8.6` soonish?

**Comment by clue** [comment]
@andig *Very soon-ish*, stay tuned :shipit:

**Comment by cboden** [comment]
Thanks @andig! Fantastic work!!! 🎉
